src/environmentd/src/http.rs

```diff
@@ -434,11 +455,23 @@ async fn internal_http_auth(mut req: Request, next: Next) -> impl IntoResponse {
 impl Server for InternalHttpServer {
     const NAME: &'static str = "internal_http";
 
-    fn handle_connection(&self, conn: Connection) -> ConnectionHandler {
+    fn handle_connection(&self, mut conn: Connection) -> ConnectionHandler {
         let router = self.router.clone();
-        let service = hyper::service::service_fn(move |req| router.clone().call(req));
 
         Box::pin(async {
+            let mut make_tower_svc = router.into_make_service_with_connect_info::<SocketAddr>();
+            let direct_peer_addr = conn.peer_addr().unwrap();
+            let peer_addr = conn
+                .take_proxy_header()
+                .await
+                .map(|h| {
+                    h.proxied_address()
+                        .map(|a| a.source)
+                        .unwrap_or(direct_peer_addr)
+                })
+                .unwrap_or(direct_peer_addr);
+            let tower_svc = make_tower_svc.call(peer_addr).await.unwrap();
+            let service = hyper::service::service_fn(|req| tower_svc.clone().call(req));
```

Comment:
Nice work stringing this all together; the `MakeService` stuff is always tricky.